src/array/cpu/csr_mm.cc

```diff
@@ -0,0 +1,104 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file array/cpu/csr_mm.cc
+ * \brief CSR Matrix Multiplication
+ */
+
+#include <dgl/array.h>
+#include <parallel_hashmap/phmap.h>
+#include <vector>
+#include "array_utils.h"
+
+namespace dgl {
+
+using dgl::runtime::NDArray;
+
+namespace aten {
+
+template <int XPU, typename IdType, typename DType>
+std::pair<CSRMatrix, NDArray> CSRMM(
+    const CSRMatrix& A,
+    NDArray A_weights,
+    const CSRMatrix& B,
+    NDArray B_weights) {
+  CHECK_EQ(A.num_cols, B.num_rows) << "A's number of columns must equal to B's number of rows";
+  const bool A_has_eid = !IsNullArray(A.data);
+  const bool B_has_eid = !IsNullArray(B.data);
+  const IdType* A_indptr = A.indptr.Ptr<IdType>();
+  const IdType* A_indices = A.indices.Ptr<IdType>();
+  const IdType* A_eids = A_has_eid ? A.data.Ptr<IdType>() : nullptr;
+  const IdType* B_indptr = B.indptr.Ptr<IdType>();
+  const IdType* B_indices = B.indices.Ptr<IdType>();
+  const IdType* B_eids = B_has_eid ? B.data.Ptr<IdType>() : nullptr;
+  const DType* A_data = A_weights.Ptr<DType>();
+  const DType* B_data = B_weights.Ptr<DType>();
+  const int64_t M = A.num_rows;
+  const int64_t P = B.num_cols;
+
+  IdArray C_indptr = IdArray::Empty({M + 1}, A.indptr->dtype, A.indptr->ctx);
+  IdType* C_indptr_data = C_indptr.Ptr<IdType>();
+  IdType nnz = 0;
+
+  // Compute indptr first in parallel
+  phmap::flat_hash_set<IdType> set;
+#pragma omp parallel for firstprivate(set)
+  for (IdType i = 0; i < M; ++i) {
+    set.clear();
+    for (IdType u = A_indptr[i]; u < A_indptr[i + 1]; ++u) {
+      IdType w = A_indices[u];
+      for (IdType v = B_indptr[w]; v < B_indptr[w + 1]; ++v)
+        set.insert(B_indices[v]);
+    }
+    C_indptr_data[i] = set.size();
+  }
+
+  IdType len = 0;
+  for (IdType i = 0; i < M; ++i) {
+    len = C_indptr_data[i];
+    C_indptr_data[i] = nnz;
+    nnz += len;
+  }
+  C_indptr_data[M] = nnz;
+
+  IdArray C_indices = IdArray::Empty({nnz}, A.indices->dtype, A.indices->ctx);
+  NDArray C_weights = NDArray::Empty({nnz}, A_weights->dtype, A_weights->ctx);
+  IdType* C_indices_data = C_indices.Ptr<IdType>();
+  DType* C_weights_data = C_weights.Ptr<DType>();
+
+  // Compute indices and data in parallel
+  phmap::flat_hash_map<IdType, DType> map;
+#pragma omp parallel for firstprivate(map)
```

Comment:
Does `firstprivate` mean `map` would be threadlocal?

Reply:
It should be thread-local and also initialized before entering the parallel region (unlike `private`).

https://stackoverflow.com/questions/15304760/how-are-firstprivate-and-lastprivate-different-than-private-clauses-in-openmp